sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.Tests/SymbolicExecution/SymbolicValues/SymbolicValue_TrySetConstraint.cs

```diff
@@ -0,0 +1,249 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using SonarAnalyzer.SymbolicExecution.Constraints;
+using Xunit;
+
+namespace SonarAnalyzer.SymbolicExecution.SymbolicValues
+{
+    public class SymbolicValue_TrySetConstraint
+    {
+        [Theory]
+        [MemberData(nameof(TrueConstraintData))]
+        [MemberData(nameof(FalseConstraintData))]
+        [MemberData(nameof(NullConstraintData))]
+        [MemberData(nameof(NotNullConstraintData))]
+        [MemberData(nameof(NoValueConstraintData))]
+        [MemberData(nameof(HasValueConstraintData))]
+        public void TrySetConstraint(SymbolicValueConstraint constraint,
+            IList<SymbolicValueConstraint> existingConstraints,
+            IList<IList<SymbolicValueConstraint>> expectedConstraintsPerProgramState)
+        {
+            // Arrange
+            var sv = new SymbolicValue();
+            var ps = SetupProgramState(sv, existingConstraints);
+
+            // Act
+            var programStates = sv.TrySetConstraint(constraint, ps).ToList();
+
+            // Assert
+            programStates.Should().HaveCount(expectedConstraintsPerProgramState.Count);
+
+            for (int i = 0; i < programStates.Count; i++)
+            {
+                var programState = programStates[i];
+                var expectedConstraints = expectedConstraintsPerProgramState[i];
+
+                foreach (var expectedConstraint in expectedConstraints)
+                {
+                    programState.HasConstraint(sv, expectedConstraint).Should().BeTrue(
+                        $"{expectedConstraint} should be present in returned ProgramState.");
+                }
+            }
+        }
+
+        public static IEnumerable<object[]> TrueConstraintData = new[]
+        {
+            new object[]

```

Comment:
I would create a small object for this instead of using object just for sake of clarity (also to be a bit more bullet proof).

Reply:
Perhaps we can experiment how to make it better later, I don't want to spend too much time on this PR; The tests that I wrote for SymbolicValue.TrySetValue are not extensive and there are many more to be written.